openshift/topics/advanced_concepts.adoc

```diff
@@ -282,12 +282,76 @@ route "sso" patched
 ----
 
 [[sso-connecting-to-an-external-database]]
-==== Connecting to an external database
+=== Connecting to an external database
 
 {project_name} can be configured to connect to an external (to OpenShift cluster) database. In order to achieve this, you need to modify the `sso-{database name}` Endpoints object to point to the proper address. The procedure is described in the https://docs.openshift.com/container-platform/latest/networking/configuring-ingress-cluster-traffic/configuring-ingress-cluster-traffic-service-external-ip.html#nw-service-external-ip_configuring-ingress-cluster-traffic-service-external-ip[OpenShift manual].
 
 Tip: The easiest way to get started is to deploy {project_name} from a template and then modify the Endpoints object. You might also need to update some of the datasource configuration variables in the DeploymentConfig. Once you're done, just roll a new deployment out.
 
+[[sso-using-custom-jdbc-driver]]
+=== Using Custom JDBC Driver
+
+To connect to any database, the JDBC driver for that database must be present and {project_name} configured
+properly. Currently, the only JDBC driver available in the image is the PostgreSQL JDBC driver. For any other
+database, you need to extend the {project_name} image with a custom JDBC driver and a CLI script to register
+it and set up the connection properties. The following steps illustrate how to do that, taking MariaDB driver
+as an example. Update the example for other database drivers accordingly.
+
+. Create an empty directory.
+. Download the JDBC driver binaries into this directory.
+. Create a new `Dockerfile` file in this directory with the following contents. For other databases, replace `mariadb-java-client-2.5.4.jar` with the filename of the respective driver:
++
+[source,subs="attributes+,macros+,+quotes"]
+----
+*FROM* {openshift_image}:latest
+
+*COPY* sso-extensions.cli /opt/eap/extensions/
+*COPY* mariadb-java-client-2.5.4.jar /opt/eap/extensions/jdbc-driver.jar
+----
+. Create a new `sso-extensions.cli` file in this directory with the following contents. Update the values of the variables in italics according to the deployment needs:
++
+[source,bash,subs="attributes+,macros+,+quotes"]
+----
+batch
+
+set DB_DRIVER_NAME=_mariadb_
+set DB_USER=_username_
+set DB_PASSWORD=_password_
+set DB_DRIVER=_org.mariadb.jdbc.Driver_
+set DB_XA_DRIVER=_org.mariadb.jdbc.MariaDbDataSource_
+set DB_JDBC_URL=_jdbc:mariadb://jdbc-host/keycloak_
+set DB_EAP_MODULE=_org.mariadb_
+
+set FILE=/opt/eap/extensions/jdbc-driver.jar
+
+module add --name=$DB_EAP_MODULE --resources=$FILE --dependencies=javax.api,javax.resource.api
```

Comment:
@hmlnarik Shouldnt' there be some line continuation character in the command?

This is what I am using:
```
[iankko@localhost custom_driver]$ cat sso-extensions.cli 
batch

set DB_DRIVER_NAME=mariadb
set DB_USERNAME=keycloak
set DB_PASSWORD=password
set DB_DRIVER=org.mariadb.jdbc.Driver
set DB_XA_DRIVER=org.mariadb.jdbc.MariaDbDataSource
set DB_JDBC_URL=jdbc:mariadb://localhost/keycloak
set DB_EAP_MODULE=org.mariadb

set FILE=/opt/eap/extensions/jdbc-driver.jar

module add --name=$DB_EAP_MODULE --resources=$FILE --dependencies=javax.api,javax.resource.api
/subsystem=datasources/jdbc-driver=$DB_DRIVER_NAME:add(
  driver-name=$DB_DRIVER_NAME,
  driver-module-name=$DB_EAP_MODULE,
  driver-class-name=$DB_DRIVER,
  driver-xa-datasource-class-name=$DB_XA_DRIVER
)

/subsystem=datasources/data-source=KeycloakDS:remove()
/subsystem=datasources/data-source=KeycloakDS:add(
  jndi-name=java:jboss/datasources/KeycloakDS,
  enabled=true,
  use-java-context=true,
  connection-url=$DB_JDBC_URL,
  driver-name=$DB_DRIVER_NAME,
  user-name=$DB_USER,
  password=$DB_PASSWORD
)

run-batch
```

And this is what I am getting in the pod log:
```
...
[0m[0m16:29:35,594 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: Red Hat Single Sign-On 7.4.0.GA (WildFly Core 10.1.2.Final-redhat-00001) started in 3822ms - Started 55 of 86 services (40 services are lazy, passive or on-demand)
[0m[[;94;109mstandalone[0m@[;94;109membedded [0m/] batch
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] 
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] set DB_DRIVER_NAME=mariadb
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] set DB_USERNAME=keycloak
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] set DB_PASSWORD=password
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] set DB_DRIVER=org.mariadb.jdbc.Driver
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] set DB_XA_DRIVER=org.mariadb.jdbc.MariaDbDataSource
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] set DB_JDBC_URL=jdbc:mariadb://localhost/keycloak
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] set DB_EAP_MODULE=org.mariadb
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] 
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] set FILE=/opt/eap/extensions/jdbc-driver.jar
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] 
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] module add --name=$DB_EAP_MODULE --resources=$FILE --dependencies=javax.api,javax.resource.api
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] /subsystem=datasources/jdbc-driver=$DB_DRIVER_NAME:add(
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   driver-name=$DB_DRIVER_NAME,
[1;91;109m'driver-name=mariadb,' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   driver-module-name=$DB_EAP_MODULE,
[1;91;109m'driver-module-name=org.mariadb,' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   driver-class-name=$DB_DRIVER,
[1;91;109m'driver-class-name=org.mariadb.jdbc.Driver,' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   driver-xa-datasource-class-name=$DB_XA_DRIVER
[1;91;109m'driver-xa-datasource-class-name=org.mariadb.jdbc.MariaDbDataSource' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] )
[1;91;109m')' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] 
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] /subsystem=datasources/data-source=KeycloakDS:remove()
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] /subsystem=datasources/data-source=KeycloakDS:add(
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   jndi-name=java:jboss/datasources/KeycloakDS,
[1;91;109m'jndi-name=java:jboss/datasources/KeycloakDS,' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   enabled=true,
[1;91;109m'enabled=true,' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   use-java-context=true,
[1;91;109m'use-java-context=true,' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   connection-url=$DB_JDBC_URL,
[1;91;109m'connection-url=jdbc:mariadb://localhost/keycloak,' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   driver-name=$DB_DRIVER_NAME,
[1;91;109m'driver-name=mariadb,' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   user-name=$DB_USER,
[1;91;109mUnrecognized variable DB_USER[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m]   password=$DB_PASSWORD
[1;91;109m'password=password' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] )
[1;91;109m')' is not a valid operation name.[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] 
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] run-batch
[1;91;109mThe batch failed with the following error (you are remaining in the batch editing mode to have a chance to correct the error): 
WFLYCTL0062: Composite operation failed and was rolled back. Steps that failed:
Step: step-1
Operation: /subsystem=datasources/jdbc-driver=mariadb:add(
Failure: WFLYCTL0155: 'driver-module-name' may not be null
[0m
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] 
[[;94;109mstandalone[0m@[;94;109membedded [0m/[;92;109m #[0m] quit
[0m16:29:35,828 INFO  [org.jboss.as] (MSC service thread 1-2) WFLYSRV0050: Red Hat Single Sign-On 7.4.0.GA (WildFly Core 10.1.2.Final-redhat-00001) stopped in 54ms
...
```

Reply:
Thanks, this is really docs issue, has been fixed in the following commit.

Reply:
Thanks, confirmed the updated version works:

![7 4_jpa_connection_details](https://user-images.githubusercontent.com/8414918/77682145-bcb55780-6f96-11ea-9c50-30b1ec279a91.png)

![7 4_demo_image_server_info](https://user-images.githubusercontent.com/8414918/77682267-f25a4080-6f96-11ea-87c2-24c160a91468.png)